library/core/src/mem/maybe_uninit.rs

```diff
@@ -172,11 +172,42 @@ use crate::ptr;
 ///
 /// ## Initializing a struct field-by-field
 ///
-/// There is currently no supported way to create a raw pointer or reference
-/// to a field of a struct inside `MaybeUninit<Struct>`. That means it is not possible
-/// to create a struct by calling `MaybeUninit::uninit::<Struct>()` and then writing
-/// to its fields.
+/// You can use `MaybeUninit<T>`, and the [`std::ptr::addr_of_mut`] macro, to initialize structs field by field:
 ///
+/// ```rust
+/// use std::mem::MaybeUninit;
+/// use std::ptr::addr_of_mut;
+///
+/// #[derive(Debug, PartialEq)]
+/// pub struct Foo {
+///     name: String,
+///     list: Vec<u8>,
+/// }
+///
+/// let foo = {
+///     let mut uninit: MaybeUninit<Foo> = MaybeUninit::uninit();
+///     let ptr = uninit.as_mut_ptr();
+///
+///     // Initializing the `name` field
+///     unsafe { addr_of_mut!((*ptr).name).write("Bob".to_string()); }
+///
+///     // Initializing the `list` field
+///     // If there was a panic here, then the `String` in the `name` field would be leaked.
```

Comment:
Language nitpick: `would be leaked` --> `leaks`

Reply:
Maybe the leak can be avoided by reordering the code a bit?

```rust
let name = ...
let list = ...

unsafe {
   addr_of_mut!((*ptr).name).write(name);
   addr_of_mut!((*ptr).list).write(list);
}
```

Reply:
@the8472 In this case yes, but I wanted this to translate to how one would initialize a large struct field by field.
If the struct had been a heap allocated struct with a large array field instead of a `Vec`, it wouldn't be possible to construct the array on the stack before moving it to the field in debug builds, since it would overflow the stack.

Reply:
Well, currently there is no guarantee that passing a large array directly to function arguments gets optimized any better than first assigning it to a named variable. You're at the mercy of the optimizer in either case since there's no placement API. https://github.com/rust-lang/rfcs/pull/2884 is still under discussion.

I think to initialize an large array without blowing the stack in debug mode you would have to loop over each item.